<!-- Launch primitive extraction from point cloud and visualization -->
<launch>

    <!-- The topic of the input cloud to segment into primitives -->
    <arg name="cloud" default="/local_metric_map/merged_point_cloud"/>
    <!-- The topic to publish the resulting found primitives on -->
    <arg name="primitive_topic" default="/plane_extraction/primitives"/>
    <!-- Choose which types of primitives to extract -->
    <arg name="extract_planes" default="true"/>
    <arg name="extract_cylinders" default="false"/>
    <arg name="extract_spheres" default="false"/>
    <!-- The min and max height between which to extract, 0 for no bounds -->
    <arg name="min_height" default="0.0"/>
    <arg name="max_height" default="0.0"/>

    <!-- Launch the primitive extraction -->
	<node pkg="primitive_extraction" type="extraction" name="plane_extraction" output="screen">
	    <!-- The min and max height between which to extract, 0 for no bounds -->
	    <param name="min_height" type="double" value="$(arg min_height)"/>
	    <param name="max_height" type="double" value="$(arg max_height)"/>
	    <param name="subsampling_voxel_size" type="double" value="0.04"/>
	    <param name="number_disjoint_subsets" type="int" value="10"/>
	    <!-- The size of the octree nodes, mainly influences speed -->
	    <param name="octree_leaf_size" type="double" value="1.0"/>
	    <!-- The size of the neighbourhood used to compute normals -->
	    <param name="normal_neigbourhood" type="double" value="0.05"/>
	    <!-- Largest distance from primitive for a point to be inlier -->
	    <param name="inlier_threshold" type="double" value="0.07"/>
	    <!-- The largest allowed deviation of point normal from surface normal -->
	    <param name="angle_threshold" type="double" value="0.3"/>
	    <!-- The probability of extracting a primitive which is not the largest -->
	    <param name="error_add_probability" type="double" value="0.001"/>
	    <!-- Do not consider any shapes with fewer inliers than this -->
	    <param name="min_inliers" type="int" value="800"/>
	    <!-- Terminate when no shapes larger than this are left -->
	    <param name="min_terminate" type="int" value="800"/>
	    <!-- Discretization of connectedness on primitive surfaces -->
	    <!-- Used to extract the largest connected region -->
	    <param name="connectedness_dist" type="double" value="0.05"/>
	    <!-- Do not consider any points further away from the camera than this -->
	    <param name="distance_threshold" type="double" value="0.0"/>
	    <!-- Choose which types of primitives to extract -->
	    <param name="extract_planes" type="bool" value="$(arg extract_planes)"/>
	    <param name="extract_cylinders" type="bool" value="$(arg extract_cylinders)"/>
	    <param name="extract_spheres" type="bool" value="$(arg extract_spheres)"/>
	    <param name="input" type="string" value="$(arg cloud)"/>
	    <param name="output" type="string" value="$(arg primitive_topic)"/>
    </node>
    
    <!-- Visualize the extracted primitives with a marker array -->
    <node pkg="primitive_extraction" type="primitives_to_markers" name="planes_to_markers" output="screen">
        <param name="input" type="string" value="$(arg primitive_topic)"/>
        <param name="output" type="string" value="/plane_extraction/primitive_marker_array"/>
        <param name="display_plane_box" type="bool" value="false"/>
    </node>

</launch>
